### PR TITLE
Be sure to detach one-shot timer watcher

### PR DESCRIPTION
**Pull request by okkez**
See https://github.com/fluent/fluentd/pull/1059/files#r75269945

### CODE REVIEW

File: lib/fluent/plugin_helper/timer.rb [hunk 1]
```diff
@@ -76,8 +76,12 @@ def on_timer
         rescue => e
           @log.error "Unexpected error raised. Stopping the timer.", title: @title, error: e
           @log.error_backtrace
-          self.detach
+          detach
           @log.error "Timer detached.", title: @title
+        ensure
+          if attached?
+            detach unless @repeating
```
Comment (tagomoris) [inline lib/fluent/plugin_helper/timer.rb]: Is `@repeating` an instance variable of `TimerWatcher`?
I think it's too internal to use it, and it looks too fragile.
We should store `repeat` into our own instance variable in `#initialize` and use it.

**Comment by okkez** [comment]
I've pushed change for comment @tagomoris

File: test/plugin_helper/test_timer.rb [hunk 2]
```diff
@@ -97,4 +97,30 @@ class Dummy < Fluent::Plugin::TestBase
...
+    counter = 0
+    d1.timer_execute(:test, 1, repeat: false) do
+      counter += 1
+    end
+
+    watchers = d1._event_loop.watchers.reject {|w| w.is_a?(Fluent::PluginHelper::EventLoop::DefaultWatcher) }
+    assert_equal(1, watchers.size)
+    assert(watchers.first.attached?)
+
+    sleep 3
```
Comment (tagomoris) [inline test/plugin_helper/test_timer.rb]: This `sleep 3` makes tests slow and unstable.
The better way to make it well is like this (not tested code):
```ruby
waiting_assertion = true
waiting_timer = true

counter = 0
d1.timer_execute(:test, 1, repeat: false) do
  sleep 0.1 while waiting_assertion
  counter += 1
  waiting_timer = false
end

# assertions about watchers here
waiting_assertion = false

sleep 0.1 while waiting_timer

assert_equal(1, counter)
# rest of assertions
```

Comment (okkez) [inline test/plugin_helper/test_timer.rb]: Thank you for sample code.
I will fix test code later.

Comment (okkez) [inline test/plugin_helper/test_timer.rb]: I've pushed a change for this comment.

**Comment by tagomoris** [comment]
LGTM.

**Comment by okkez** [comment]
Thanks!

**Pull request merged**